Port overlays documentation (#171)
microsoft/vcpkg-docs: vcpkg/users/port-overlays.md

---
title: Port overlays
description: This article describes overlay ports in vcpkg. Overlay ports are used to force a specific port definition to be used by vcpkg during the package installation process.
author: data-queue
ms.author: danshaw2
ms.date: 9/20/2023
ms.prod: vcpkg

> ms.topic: conceptual

Reviewer: ```suggestion
ms.topic: concept-article
```

After reading the classifications once more, I think this should be [`concept-article`](https://review.learn.microsoft.com/help/platform/metadata-taxonomies#mstopic):

![image](https://github.com/microsoft/vcpkg-docs/assets/3526662/6323f05e-f47c-47bd-95ab-209f05c7ffa4)

---

# Overlay ports

> ## Overview

Reviewer 2: Our existing "concept" documentation doesn't include an ## Overview section
```suggestion

```

An overlay port can act as a drop-in replacement for an existing port. An overlay port can also add a new port that is otherwise not available in a [registry](../maintainers/registries.md). While [resolving package names](registries.md#package-name-resolution), overlay ports take priority, forcing overlay ports to be selected.

> Naming conflicts can arise when multiple overlays introduce the same port. For instance, overlay directories `[dirA, dirB]` may both introduce a port with the same name: `dirA/zlib` and `dirB/zlib`. vcpkg resolves `zlib` in order of which directory is specified first. In this example, `dirA/zlib` is selected as the `zlib` port. If overlay port options are specified in multiple places, overlay resolution priority follows first from the command line, the manifest, and finally the environment.

Reviewer 2: ```suggestion
Naming conflicts can arise when multiple overlays introduce the same port. For instance, overlay directories `[dirA, dirB]` may introduce a port with the same name: `dirA/zlib` and `dirB/zlib`. vcpkg resolves `zlib` in order of which directory is specified first. This example selects `dirA/zlib` as the `zlib` port. If overlay port options are specified in multiple places, overlay resolution priority follows first from the command line, then the manifest, and finally, the environment.
```

> ## Configuration Syntax

Reviewer 2: ```suggestion
## Using an overlay port
```
?

Do we have docs for creating an overlay port? That way, we can describe it and link to a tutorial for creating an overlay port. I ask because the crux of this article is describing how to use an overlay port.

Reviewer: Yes, we have an article "Use a locally modified dependency" (or some similar title) which is intended to be a 1-2 combo of applying patches and using overlays tutorials.

Author: fixed

A directory refers to a set of overlay ports and can be specified as follows: 
- Single overlay port: `<directory>/sqlite3` refers to a single port
- Directory of overlay ports: `<directory>` refers to a directory of ports

> A valid port must contain both `vcpkg.json` and `portfile.cmake`.

Author: I think it is more syntax related, so I've kept it in the current section

You can add an overlay port in several ways:
* Command-line: Add one or multiple `--overlay-ports=<directory>` options to your vcpkg command
* [Manifest](../reference/vcpkg-configuration-json.md#overlay-ports): Populate the `"overlay-ports"` array in `vcpkg-configuration.json`
* [Environmental variable](./config-environment.md#vcpkg_overlay_ports): Set `VCPKG_OVERLAY_PORTS` to a list of directory paths

### Example: Overlay Ports Example

Given this directory structure:
```
team-ports/
|-- sqlite3/
|---- vcpkg.json
|---- portfile.cmake
|-- rapidjson/
|---- vcpkg.json
|---- portfile.cmake
|-- curl/
|---- vcpkg.json
|---- portfile.cmake

my-ports/
|-- sqlite3/
|---- vcpkg.json
|---- portfile.cmake
|-- rapidjson/
|---- vcpkg.json
|---- portfile.cmake

vcpkg
|-- ports/
|---- <upstream ports>
|-- vcpkg.exe

> ```

Reviewer: I'm concerned about how screen readers will treat this section. We should replace it with an image with appropriate description and alt-text.

> Running:

Reviewer: ```suggestion
When run:
```
Machine translation doesn't do well with gerunds.

```
vcpkg.exe install sqlite3 --overlay-ports=my-ports --overlay-ports=team-ports

> ```

Reviewer: Set the code-block language to PowerShell or Bash, since this is a conceptual article, we don't need to provide instructions for each platform, we just pick one. Also, remove the `.exe` from the command.

---
title: Port overlays
description: Port overlays
author: data-queue
ms.author: danshaw2
ms.date: 9/20/2023
ms.prod: vcpkg
ms.topic: conceptual
---

# Port Overlays

## Overview
The default catalog in vcpkg comprises libraries maintained by both the vcpkg team and users. The catalog does not contain the entire universe of libraries, every possible version or patch.

To facilitate the usage of these missing libraries without having to either add them to the default catalog or manage a custom registry, vcpkg uses port overlays. Port overlays are effective when you don't want to worry about versioning and simply want to use a library.

## Overlay Ports

> An overlay port typically refers to a new port or a drop-in replacement for an existing port. In either case, the overlay port takes priority in vcpkg dependency resolution, ensuring it gets selected. However, a potential pitfall is the naming conflict that can arise when multiple overlays introduce the same port.

Reviewer: ```suggestion
An overlay port can act as a drop-in replacement for an existing port. An overlay port can also add a new port that is otherwise not available in a [registry](../maintainers/registries.md). While [resolving package names](registries.md#package-name-resolution), overlay ports take priority, forcing the overlay ports to be selected. 
```

> For instance, `overlay-dir-A` and `overlay-dir-B` may both introduce the same overlay port `zlib`. In this situation, vcpkg resolves `zlib` based on the lookup order in the list of overlay port paths.

Reviewer: We should elaborate on this example and explain the rules of overlay port resolution priority.

Author: good point, I've added some more context

## Configuration Syntax

A directory refers to a set of overlay ports and can be specified as follows: 
- Single overlay port: `<directory>/sqlite3` refers to a single port
- Directory of overlay ports: `<directory>` refers to a directory of ports

> A valid port must contain both `vcpkg.json` and `portfile.cmake`.

Reviewer: In general, this paragraph needs to be rewritten to look closer to our current documentation style. Also, since this section is called `Configuration Syntax` it should limit to explain the syntax and not delve on the specifics of what the "overlay ports" directories should contain.

Author: I noticed we talked about overlay ports priority resolution in the Registries concepts documentation. I wonder if we should duplicate that here. Thoughts?
